src/pushsource/_impl/reader/base.py

```diff
@@ -0,0 +1,129 @@
+import logging
+
+from io import BufferedIOBase
+from urllib import parse
+
+LOG = logging.getLogger("pushsource")
+
+
+class ContentReader(BufferedIOBase):
```

Comment:
I have some issues with this class:

- Firstly I think the concept of `register_reader` and `get` is not needed.  The issue didn't ask for anything like this.  I would expect in most cases whatever `Source` backend is providing a push item knows exactly what reader makes sense, and doesn't benefit from this additional layer of abstraction.
- I think the concept of a *base class* for all readers doesn't really work either.  Again looking at the idea of doing an HTTP request as an example: the HTTP client API is probably going to itself give you a file-like object.  You can't make the client return some subclass of ContentReader.  All you can do is accept the object it gives you, and wrap it in something else if you need to.  So, this would seem to point us towards composition (rather than inheritance) being the right approach here.

Considering what this class actually needs to do beyond the standard file-like object API, I believe there are just two points requested:

- it should have a `name` attribute (str)
- it should not be seekable

So, I think the better design would be to have a class not designed to be subclassed, but designed to *wrap* other io objects and enforce these two behaviors.

`io.BufferedReader` is already designed to wrap another stream, so I think we could just extend that slightly to add a `name` attribute and enforce non-seekability.

Something like this:

```python
class PushItemReader(BufferedReader):
    # A BufferedReader which always has a name and is always unseekable.
    def __init__(self, raw, *, name, buffer_size=DEFAULT_BUFFER_SIZE):
        self.name = name
        super(raw, buffer_size)

    def seekable(self):
        return False

    def seek(self, pos, whence=0):
        raise UnsupportedOperation(f"Seek unsupported while reading {self.name}")
```

This would be a private class, not exposed or documented as public API.

Then the way it'd be used internally is something like this:

```python
# Example of simple 'open file' case
def content(self):
    return PushItemReader(open(self.src), name=self.src)

# (possible future) example of HTTP case
def content(self):
    r = requests.get(self.src, stream=True)
    check_success(r)
    return PushItemReader(r.raw, name=self.src)
```

Reply:
Thank you for sharing this helpful insight. It seems I over-analysed and complicated it as you mentioned.

I have the understanding that the `Source` backend gets the list and the metadata of the push items. There might be a case that the push items reside somewhere else in a separate artifact repository. 
Hence , `Source` backend only knows the artifact repo and returns the corresponding `PushItem` with the the metadata. e.g. Errata Tool provides the push item metadata that may reside in some object store like S3, some other location reachable via HTTP etc. or a mix of them depending on the push item types or creation time like old ones on brew and new ones one some other artifact store. So `ErrataSource` backend with provide the push items accordingly without getting into how to fetch them.

I thought it's the `PushItem` that should responsible to gets the actual content from the artifact repo using the applicable `PushItemReader`. `PushItem` delegates all the responsibility to the `PushItemReader` to get the content including getting the raw IO object, wrapping it and any other steps like auth, config etc. that are required to get the content from the artifact repo. This could decouple the entire content fetching responsibility from the `Source` or `PushItem`

Eventually, if there a new artifact location OR a new workflow involving a new backend or push item type, a new `PushItemReader` will be registered to be used by the corresponding `PushItem`.

Extending `io.BufferedReader` in `PushItemReader` and wrapping the raw IO object definitely helps and doesn't require this `ContentReader` class. However, what do you think about extending the `PushItemReader` for specific content source or content type?

Reply:
I haven't fully worked through the idea of how exactly to make the `content`/`open` method actually do something other than `open(self.src)`.  I would rather not commit to anything until it comes time to really implement an alternative method of opening, as I see this issue as being primarily about allowing *the users of pushsource* to prepare for alternative methods of opening.  We need to change the interface to achieve that, but implementation details can be left for later.

But anyway, here's a concrete idea which I would use as a starting point.

The `PushItem` class can have an 'opener' attribute, something like this:

```
opener = attr.ib(default=open_src, repr=False)
```

(Originally I was thinking this should probably be private, but since we do allow `Source` implementations to be registered from outside of pushsource library, I guess it should actually be public.)

The default implementation would be something like:

```
def open_src_local(item):
    # default opener for push items, assumes that the item's
    # 'src' points at a locally-accessible file
    return open(item.src, 'rb')
```

Then a `Source` implementation can change the method used to open by passing a value during push item creation.  For example, in `KojiSource` where RpmPushItems are currently instantiated:

```
        return [
            RpmPushItem(
                name=os.path.basename(rpm_path),
                src=rpm_path,
                dest=self._dest,
                signing_key=rpm_signing_key,
                build=build["nvr"],
            )
        ]
```

The code could potentially be updated to do something like:

```
        kwargs = dict(src=rpm_path)
        # We still default to reading files locally, but when koji source was constructed,
        # was an argument passed to request downloading over HTTP?
        if self._download_files:
          rpm_url = os.path.join(koji_download_root, rpm_path)
          kwargs['src'] = rpm_url
          kwargs['opener'] = utils.open_src_http

        return [
            RpmPushItem(
                name=os.path.basename(rpm_path),
                dest=self._dest,
                signing_key=rpm_signing_key,
                build=build["nvr"],
                **kwargs,
            )
        ]
```

Still I would not recommend adding anything like this now (except perhaps as a local experiment).
This change is first and foremost about making it possible for the users of this library to transfer the responsibility for reading content into the `pushsource` library, this does not yet require us to make all the design decisions related to supporting alternative methods of reading.